Mathlib/NumberTheory/FLT/Polynomial.lean

```diff
@@ -0,0 +1,332 @@
+    rw [natDegree_mul (mul_ne_zero ha hb) hc,
+        natDegree_mul ha hb, Nat.add_one_le_iff] at hpa hqb hrc
+
+    exfalso
+    exact (ineq_pqr_contradiction hp hq hr hineq hpa hqb hrc)
+  · rw [derivative_C_mul, derivative_C_mul, derivative_C_mul,
+        mul_eq_zero_left_iff (C_ne_zero.mpr hu),
+        mul_eq_zero_left_iff (C_ne_zero.mpr hv),
+        mul_eq_zero_left_iff (C_ne_zero.mpr hw),
+        derivative_pow_eq_zero_iff chp,
+        derivative_pow_eq_zero_iff chq,
+        derivative_pow_eq_zero_iff chr] at dr0
+    exact dr0
+
+-- helper lemma that gives a baggage of small facts on `contract (ringChar k) a`
+private lemma find_contract {a : k[X]}
+    (ha : a ≠ 0) (hda : derivative a = 0) (chn0 : ringChar k ≠ 0) :
+    ∃ ca, ca ≠ 0 ∧
+      a = expand k (ringChar k) ca ∧
+      a.natDegree = ca.natDegree * ringChar k := by
+  have heq := (expand_contract (ringChar k) hda chn0).symm
+  refine ⟨contract (ringChar k) a, ?_, heq, ?_⟩
+  · intro h
+    rw [h, map_zero] at heq
+    exact ha heq
+  · rw [← natDegree_expand, ← heq]
+
+private theorem expand_dvd {a b : k[X]} (n : ℕ) (h : a ∣ b) :
+    expand k n a ∣ expand k n b := by
+  rcases h with ⟨t, eqn⟩
+  use expand k n t
+  rw [eqn, map_mul]
+
+variable [DecidableEq k]
+
+private theorem is_coprime_of_expand
+    {a b : k[X]} {n : ℕ} (hn : n ≠ 0) :
+    IsCoprime (expand k n a) (expand k n b) → IsCoprime a b := by
+  simp_rw [← EuclideanDomain.gcd_isUnit_iff]
+  cases' EuclideanDomain.gcd_dvd a b with ha hb
+  have he := EuclideanDomain.dvd_gcd (expand_dvd n ha) (expand_dvd n hb)
+  intro hu
+  have heu := isUnit_of_dvd_unit he hu
+  rw [Polynomial.isUnit_iff] at heu ⊢
+  rcases heu with ⟨r, hur, eq_r⟩
+  rw [eq_comm, expand_eq_C (zero_lt_iff.mpr hn), eq_comm] at eq_r
+  exact ⟨r, hur, eq_r⟩
+
+theorem Polynomial.flt_catalan_aux
+    {p q r : ℕ} {a b c : k[X]} {u v w : k}
+    (heq : C u * a ^ p + C v * b ^ q + C w * c ^ r = 0)
+    (hp : 0 < p) (hq : 0 < q) (hr : 0 < r)
+    (hineq : q * r + r * p + p * q ≤ p * q * r)
+    (chp : ¬ringChar k ∣ p) (chq : ¬ringChar k ∣ q) (chr : ¬ringChar k ∣ r)
+    (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0) (hab : IsCoprime a b)
+    (hu : u ≠ 0) (hv : v ≠ 0) (hw : w ≠ 0) :
+    a.natDegree = 0 := by
+  cases' eq_or_ne (ringChar k) 0 with ch0 chn0
+  -- characteristic zero
+  · have hderiv := flt_catalan_deriv
+      hp hq hr hineq chp chq chr ha hb hc hab hu hv hw heq
+    rcases hderiv with ⟨da, -, -⟩
+    have czk : CharZero k := by
+      apply charZero_of_inj_zero
+      intro n
+      rw [ringChar.spec, ch0]
+      exact zero_dvd_iff.mp
+    rw [eq_C_of_derivative_eq_zero da]
+    exact natDegree_C _
+  -- characteristic p > 0
+  · set d := a.natDegree with eq_d; clear_value d
+    -- set up infinite descent
+    -- strong induct on `d := a.natDegree`
+    revert ha hb hc hab heq
+    revert eq_d
+    revert a b c
+    induction' d using Nat.case_strong_induction_on with d ih_d
+    · intros; solve_by_elim
```

Comment:
This isn't idiomatic. Firstly I think `induction'` is now discouraged and one should use Lean 4's `induction`. Secondly `induction ... generalizing` is supposed to be used instead of the `revert...intros` hack.

Reply:
@seewoo5 Can I handle this?